hazelcast/src/main/java/com/hazelcast/collection/impl/collection/operations/CollectionMergeOperation.java

```diff
@@ -62,45 +57,31 @@ public int getId() {
 
     @Override
     public boolean shouldBackup() {
-        return valueMap != null && !valueMap.isEmpty();
+        return backupCollection != null && !backupCollection.isEmpty();
     }
 
     @Override
     public Operation getBackupOperation() {
-        return new CollectionAddAllBackupOperation(name, valueMap);
+        return new CollectionMergeBackupOperation(name, backupCollection);
     }
 
     @Override
     public void run() throws Exception {
         CollectionContainer collectionContainer = getOrCreateContainer();
-        valueMap = createHashMap(mergingValues.size());
-        for (CollectionMergeTypes mergingValue : mergingValues) {
-            CollectionItem mergedItem = collectionContainer.merge(mergingValue, mergePolicy);
-            if (mergedItem != null) {
-                valueMap.put(mergedItem.getItemId(), mergedItem.getValue());
-            }
-        }
+        backupCollection = collectionContainer.merge(mergingValue, mergePolicy);
```

Comment:
Do we need to handle the case when the result of the merge is `null` and destroy the collection then? This is how I did it for the ringbuffer.
I guess right now if the user returns `null`, the collection remains as it was on the bigger cluster.

Reply:
It's the question how we want to design it. What should happen on returning
* `null`
* an empty collection

The same goes for key-based data structures in the case we return `null`. This will need fixing in all data structures (I think only `IMap` handles this so far).

Reply:
Yes, I drew inspiration from how map handled it - https://github.com/hazelcast/hazelcast/blob/00ee8406b7945b601fc6c2e3783b7c53a5dd582d/hazelcast/src/main/java/com/hazelcast/map/merge/MapMergePolicy.java#L47.
But I saw that our new contract didn't mention it. Maybe we don't need to handle it now, especially if we mark the SPI as `Beta`

Reply:
I'll try to make this consistent now, otherwise we'll confuse everyone, including us.

`CollectionContainer.merge()`:
```java
if (existingValue == null) {
    createNewCollectionItems(existingItems, newValues, serializationService);
} else if (newValues == null || newValues.isEmpty()) {
    existingItems.clear();
} else if (!newValues.equals(existingValue.getValue())) {
    existingItems.clear();
    createNewCollectionItems(existingItems, newValues, serializationService);
}
```
This should not replace the items, when you return `existingValue.getValue()`, e.g. in the `PutIfAbsentMergePolicy`, to prevent the re-creation of the existing items.

I also tried to prevent the re-creation of the same backups in `CollectionMergeOperation`:
```java
private transient boolean currentCollectionIsEmpty;
private transient long currentCollectionItemId;
private transient Collection<CollectionItem> backupCollection;

@Override
public void run() throws Exception {
    CollectionContainer collectionContainer = getOrCreateContainer();
    currentCollectionIsEmpty = collectionContainer.getCollection().isEmpty();
    currentCollectionItemId = collectionContainer.getCurrentId();

    backupCollection = collectionContainer.merge(mergingValue, mergePolicy);
}

@Override
public boolean shouldBackup() {
    return currentCollectionIsEmpty != backupCollection.isEmpty()
            || currentCollectionItemId != getOrCreateContainer().getCurrentId();
}
```